jlm/rvsdg/gamma.cpp

```diff
@@ -265,79 +265,130 @@ GammaOperation::operator==(const operation & other) const noexcept
   {
-    argument = dynamic_cast<const rvsdg::RegionArgument *>(result(n)->origin());
-    if (argument == nullptr || argument->input()->origin() != origin)
-      break;
+    ev.branches.push_back(subregion(n)->argument(index));
   }
+  return ev;
+}
 
-  auto isInvariant = (n == nresults());
-  if (isInvariant && invariantOrigin != nullptr)
+std::vector<GammaNode::EntryVar>
+GammaNode::GetEntryVars() const
+{
+  std::vector<GammaNode::EntryVar> vars;
+  for (size_t n = 0; n < ninputs() - 1; ++n)
   {
-    *invariantOrigin = origin;
+    vars.push_back(GetEntryVar(n));
   }
-
-  return isInvariant;
+  return vars;
 }
 
-/* gamma node */
+GammaNode::EntryVar
+GammaNode::MapInputEntryVar(const rvsdg::input & input) const
+{
+  JLM_ASSERT(rvsdg::TryGetOwnerNode<GammaNode>(input) == this);
+  JLM_ASSERT(input.index() != 0);
+  return GetEntryVar(input.index() - 1);
+}
 
-GammaNode::~GammaNode() noexcept = default;
+GammaNode::EntryVar
+GammaNode::MapBranchArgumentEntryVar(const rvsdg::output & output) const
+{
+  JLM_ASSERT(rvsdg::TryGetRegionParentNode<GammaNode>(output) == this);
+  return GetEntryVar(output.index());
+}
 
-const GammaNode::entryvar_iterator &
-GammaNode::entryvar_iterator::operator++() noexcept
+GammaNode::ExitVar
+GammaNode::AddExitVar(std::vector<jlm::rvsdg::output *> values)
 {
-  if (input_ == nullptr)
-    return *this;
+  if (values.size() != nsubregions())
+    throw jlm::util::error("Incorrect number of values.");
 
-  auto node = input_->node();
-  auto index = input_->index();
-  if (index == node->ninputs() - 1)
+  const auto & type = values[0]->Type();
+  auto output = node::add_output(std::make_unique<rvsdg::StructuralOutput>(this, type));
+
+  std::vector<rvsdg::input *> branch_results;
+  for (size_t n = 0; n < nsubregions(); n++)
   {
-    input_ = nullptr;
-    return *this;
+    branch_results.push_back(&rvsdg::RegionResult::Create(
+        *subregion(n),
+        *values[n],
+        static_cast<rvsdg::StructuralOutput *>(output),
+        output->Type()));
   }
 
-  input_ = static_cast<GammaInput *>(node->input(++index));
-  return *this;
+  return ExitVar{ std::move(branch_results), std::move(output) };
 }
 
-const GammaNode::exitvar_iterator &
-GammaNode::exitvar_iterator::operator++() noexcept
+std::vector<GammaNode::ExitVar>
+GammaNode::GetExitVars() const
 {
-  if (output_ == nullptr)
-    return *this;
+  std::vector<GammaNode::ExitVar> vars;
+  for (size_t n = 0; n < noutputs(); ++n)
+  {
+    std::vector<rvsdg::input *> branch_results;
+    for (size_t k = 0; k < nsubregions(); ++k)
+    {
+      branch_results.push_back(subregion(k)->result(n));
+    }
+    vars.push_back(ExitVar{ std::move(branch_results), output(n) });
+  }
+  return vars;
+}
 
-  auto node = output_->node();
-  auto index = output_->index();
-  if (index == node->nexitvars() - 1)
+GammaNode::ExitVar
+GammaNode::MapOutputExitVar(const rvsdg::output & output) const
+{
+  std::vector<rvsdg::input *> branch_results;
```

Comment:
Some assert that output is the output of a gamma node.

Reply:
added assertion